CoolMasterNet Climate platform

OnFreund opened the pull request:
## Description:

[CoolMasterNet](https://coolautomation.com/products/coolmasternet/) as a Climate platform. This allows you to controller multiple HVAC instances connected to a single CoolMasterNet controller.
The integration is based on [pycoolmaster](https://github.com/koreth/pycoolmasternet).

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8422

## Example entry for `configuration.yaml` (if applicable):
```yaml
climate:
  - platform: coolmaster
    host: YOUR_COOLMASTER_HOST
    port: YOUR_COOLMASTER_PORT
    supported_modes:
      - heat
      - cool
      - dry
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

OnFreund committed:
CoolMasterNet Climate platform

OnFreund commented:
Might be missing something, but it seems like the Travis failure isn't directly related to my changes:
```
3.5.3 is not installed; attempting download
Downloading archive: https://s3.amazonaws.com/travis-python-archives/binaries/ubuntu/14.04/x86_64/python-3.5.3.tar.bz2
$ curl -sSf -o python-3.5.3.tar.bz2 ${archive_url}
curl: (56) SSL read: error:00000000:lib(0):func(0):reason(0), errno 104
Unable to download 3.5.3 archive. The archive may not exist. Please consider a different version.
```

OnFreund commented:
Another failure that doesn't seem related:
```
=================================== FAILURES ===================================
________ TestStatisticsSensor.test_initialize_from_database_with_maxage ________
```

OnFreund commented:
Hmm, not sure how to get someone to review this. I think this could be a helpful platform...

andrewsayre commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+    @property
+    def fan_list(self):
+        """Return the list of available fan modes."""
+        return self._fan_list
```
Return the list directly

andrewsayre commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+        self._operation_list = supported_modes
+        self._support_flags = SUPPORT_FLAGS
+        self._fan_list = ['low', 'med', 'high', 'auto']
+        self._target_temperature = 0
```
Use `None` for initial values.

andrewsayre commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+        self._current_fan_mode = ''
+        self._current_operation = ''
+        self._on = False
+        self._unit = default_unit
```
I don't think this is needed as initial value should be `None` -- referring to `default_unit`

andrewsayre commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+
+    def update(self):
+        """Pull state from CoolMasterNet."""
+        status = self._device.status
```
Does accessing the `status` property cause the latest data to be fetched?

OnFreund replied:
[Yes, it does.](https://github.com/koreth/pycoolmasternet/blob/36bc759d8685fd8bab5a0e65578b10064f082e34/pycoolmasternet/__init__.py#L138)

andrewsayre commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+
+DEFAULT_PORT = 10102
+
+DOMAIN = 'coolmaster'
```
Minor, but suggest moving this up under `REQUIREMENTS` so it's easier to find.

andrewsayre reviewed (changes requested):
Overall looks pretty good!  Some additional comments:
1. Do you want to add yourself to CODEOWNERS so you get tagged for future PRs over this code?
2. Would you consider adding Unit Tests (and undo the `.coveragerc` change).  Testing a platform is pretty easy and I see very little mocking needed.  See the SmartThings component for some example platform tests.

OnFreund committed:
Address Coolmaster PR comments

OnFreund commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+
+    def __init__(self, device, supported_modes, default_unit):
+        """Initialize the climate device."""
+        _LOGGER.debug("Adding device %s", device.uid)
```
Thanks. I changed the log string instead, since I have no other use for `async_added_to_hass`.

OnFreund commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,207 @@
+            self.schedule_update_ha_state()
+
+    def set_fan_mode(self, fan_mode):
+        """Set new target temperature."""
```
Wow, thanks. I guess I relied to much on the demo climate platform.

OnFreund committed:
Fix docstrings on climate demo platform

OnFreund commented:
@andrewsayre thanks so much.
I addressed all your comments (and also fixed the original wrong docstrings on the demo platform), and added myself as a code owner.

As for tests, since there's such little logic here, it felt like the added tax here for everyone else developing is not worth it, but if you feel otherwise I'm happy to reconsider.

andrewsayre commented:
> As for tests, since there's such little logic here, it felt like the added tax here for everyone else developing (the test suite takes long as it is, adding more tests makes it slower) is not worth it, but if you feel otherwise I'm happy to reconsider.

I think there's always value in adding tests--I subscribe to the philosophy, _If there aren't tests, it doesn't work_.  Certainly something to consider in the future!

CLA bot is stuck, but everything looks good.  I'll approve.  @MartinHjelmare would you take a quick pass given this is my first time reviewing a new platform? :)

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+        self._on = status['is_on']
+
+        device_mode = status['mode']
+        if device_mode == FAN_ONLY_OVERRIDE:
```
We prefer to make two dicts at the module level with the device mode mapped to the home assistant state modes and vice versa, so that it's clear what modes are used.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+        return self._uid
+
+    @property
+    def device_info(self):
```
This is only read when implementing config entries. We can remove it for now.

https://developers.home-assistant.io/docs/en/device_registry_index.html#defining-devices

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return True
```
This is the default.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+        _LOGGER.debug("Setting fan mode of %s to %s", self.unique_id,
+                      fan_mode)
+        self._device.set_fan_speed(fan_mode)
+        self.schedule_update_ha_state()
```
We can remove this. It will be called at the end of the service call automatically, since this is a polling entity.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+        _LOGGER.debug("Setting operation mode of %s to %s", self.unique_id,
+                      operation_mode)
+        self._device.set_mode(operation_mode)
+        self.schedule_update_ha_state()
```
See above.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+        """Turn on."""
+        _LOGGER.debug("Turning %s on", self.unique_id)
+        self._device.turn_on()
+        self.schedule_update_ha_state()
```
See above.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+        """Turn off."""
+        _LOGGER.debug("Turning %s off", self.unique_id)
+        self._device.turn_off()
+        self.schedule_update_ha_state()
```
See above.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+
+    def set_operation_mode(self, operation_mode):
+        """Set new operation mode."""
+        if operation_mode == STATE_FAN_ONLY:
```
See above about operation modes mapping.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+            _LOGGER.debug("Setting temp of %s to %s", self.unique_id,
+                          str(temp))
+            self._device.set_thermostat(str(temp))
+            self.schedule_update_ha_state()
```
See below.

MartinHjelmare reviewed (changes requested):
Generally looks good. Some adjustments needed.

MartinHjelmare commented on .coveragerc:
```diff
@@ -67,6 +67,7 @@ omit =
     homeassistant/components/camera/xiaomi.py
     homeassistant/components/camera/yi.py
     homeassistant/components/cast/*
+    homeassistant/components/climate/coolmaster.py
```
Eg that would be `homeassistant/components/coolmaster/climate.py`.

OnFreund replied:
I couldn't find any documentation for this. I did however see a script by @balloob that's supposed to do this in mass, and it seems like it hasn't run for climate yet since there are other platforms in there. Can we keep it like this and have the script change it when run?

MartinHjelmare replied:
Yes, it's ok to keep for now. It's just a heads up. We probably will merge this before the climate platforms are converted.

OnFreund commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+    supported_modes = config.get(CONF_SUPPORTED_MODES)
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    cool = CoolMasterNet(host, port=port)
```
Through the 3rd party, it eventually does I/O through `telnetlib.Telnet`.
The async io docs for a platform simply mention replacing `setup_platform` with its async counterpart, so I'm not sure what's the best way to handle this.

MartinHjelmare replied:
Yes, just replace `async_setup_platform` with `setup_platform`. Then the calls will be done in our thread pool.

OnFreund replied:
Got it, thanks. Will be helpful to see some reference async implementation of a platform doing basic I/O. The docs aren't too clear.

MartinHjelmare replied:
It's possible to use `hass.async_add_executor_job` to schedule calls on the thread pool from async context, but if the library isn't supporting asyncio, there's usually no point in using the home assistant async api.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+    cool = CoolMasterNet(host, port=port)
+
+    all_devices = [CoolmasterClimate(device, supported_modes)
+                   for device in cool.devices()]
```
The `cool.devices` call isn't async safe.

OnFreund commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,205 @@
+
+    def __init__(self, device, supported_modes):
+        """Initialize the climate device."""
+        _LOGGER.debug("Creating device %s", device.uid)
```
OK, moving to `setup_platform` where they are found.

OnFreund committed:
Additional CoolMaster PR review fixes

OnFreund commented:
Thanks @MartinHjelmare. I addressed all your comments.

MartinHjelmare commented on homeassistant/components/climate/coolmaster.py:
```diff
@@ -0,0 +1,186 @@
+        _LOGGER.debug("Found device %s", device.uid)
+
+    all_devices = [CoolmasterClimate(device, supported_modes)
+                   for device in devices]
```
Append the created entities to the list in the first loop after the debug log instead. We'll save one loop.

OnFreund replied:
Not sure I agree with this. Since the number of items is capped by about 200, there's not going to be any major performance improvement here, and the runtime is anyway going to be dominated by the additional loop that gets the status of the devices after they're initialized.
I find that loops that combine a map and a side effect are usually more verbose, harder to read and more error prone than two separate loops.
Is there some Python syntax I don't know that would make this better? (this is my first real piece of code with Python, so my experience is based on other languages).

OnFreund replied:
As a compromise, I separated building the entity to its own function. Now there's only one loop, but the function has a side effect.

MartinHjelmare reviewed (approved):
Nice! Can be merged when last comment is addressed.

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
For the future, please don't squash commits after review has started to make it easier for readers to track changes. We will squash all the commits when merging.
